Source/WebInspectorUI/UserInterface/Views/View.js

```diff
@@ -285,6 +285,7 @@ WI.View = class View extends WI.Object
     {
         this._setDirty(false);
         let isInitialLayout = !this._didInitialLayout;
+        let hasNonDirtyParentView = !this.parentView.layoutReason && !this.parentView.dirty;
```

Comment:
I think we also need to check the `_dirtyDescendantsCount` as that's really what determines whether this is the last child `WI.View` in need of a `layout`.  the problem with that, however, is that `_dirtyDescendantsCount` keeps track of the entire subtree, not just the direct children, so you'd have to check that after calling `layout`, but there's also nothing stopping a `WI.View` from calling `needsLayout` inside `layout`, so it's possible that even looking at `_dirtyDescendantsCount` won't be a good signal.

we might need to add a separate `_dirtyChildCount` that keeps track of just the direct children (and then maybe rename `_dirtyDescendantsCount` to be more specific to the fact that it's about grandchildren and below) and then check to see when it's equal to `1` (i.e. "this is the last dirty child") before we `this._setDirty(false)`

also please add some tests for this in `LayoutTests/inspector/view/` :)

Reply:
> I think we also need to check the `_dirtyDescendantsCount` as that's really what determines whether this is the last child `WI.View` in need of a `layout`.

I've been thinking about this and looking at uses. 

We're not trying to determine the **last** `WI.View` in need of a `layout`. We're looking for the **parent of the first** `WI.View` in need of a `layout`.

I don't think we need an additional check other than a missing `WI.View.layoutReason`. 

```diff
- let hasNonDirtyParentView = !this.parentView.layoutReason && !this.parentView.dirty;
+ let hasNonDirtyParentView = !this.parentView?.layoutReason;
```

Looking at the dirty state of the parent view when doing layout for the subview is moot because it always `false`.

Here's how I understand layout of view tree happens:

`WI.View._layoutSubtree()` gets called from one of two paths:
- directly when calling `WI.View.updateLayout()`
- indirectly when scheduling with`WI.View.needsLayout()` -->  `WI.View._scheduleLayoutForView()`

In both cases, the layout happens from top to bottom along the view tree.
When called directly, it starts from a specific view.
When scheduled, it starts from `_rootView`. 

When called directly, a `WI.View._layoutReason` is set, either explicitly or fall back to `WI.View.LayoutReason.Dirty`.
When scheduled, a dirty flag is set: `WI.View._setDirty(true)`. **Here we can also set `view._setLayoutReason(WI.View.LayoutReason.Dirty)`**. 

When layout starts from the top-most view that needs layout, the first thing that happens in `WI.View._layoutSubtree()` is to flip its own dirty flag, `this._setDirty(false)`, then call `this.layout()`, then recursively call `view._layoutSubtree()` on all its subviews (while passing down the `_layoutReason`), and **finally reset the `_layoutReason` to null**.

So by the time a subview does layout, its parent's `_layoutReason` hasn't been reset yet. 
And _all_ subviews will have `_layoutSubtree()` called on them.

From this behavior we can infer that the parent view which doesn't have a truthy `_layoutReason` was not `dirty`. That's our target parent view for which we want to call `didLayoutSubtree()`.

The modifications that I need to make are:
```diff
_layoutSubtree()
{
-   let hasNonDirtyParentView = !this.parentView.layoutReason && !this.parentView.dirty;
+   let hasNonDirtyParentView = !this.parentView?.layoutReason;
...
}

static _scheduleLayoutForView(view)
{
...
   view._setDirty(true);
+  view._setLayoutReason(WI.View.LayoutReason.Dirty);
...
```

I believe this handles most cases, including: 
> there's also nothing stopping a WI.View from calling needsLayout inside layout,
because that goes via the scheduling route, where we'll set the layout reason.

Is there any use case I'm missing from this logic?
If it works, I prefer this over keeping counts of child views.

Reply:
> We're not trying to determine the **last** `WI.View` in need of a `layout`. We're looking for the **parent of the first** `WI.View` in need of a `layout`.

But what about if two children of the same `WI.View` are dirty?  We only want to call `didLayoutSubtree` after all children of a given `WI.View` are done with their `layout` (for a single `updateLayout`/`WI.View._visitViewTreeForLayout`).  Calling `didLayoutSubtree` after the first child `WI.View` means that the parent could be looking at incorrect state for any of its other child `WI.View` that would then update when they do their `layout`.

Consider this example

```
WI.CounterView = class CounterView extends WI.View
{
    layout()
    {
        this.count = (this.count || 0) + 1;
    }
};

WI.ParentView = class ParentView extends WI.View
{
    initialLayout()
    {
        this.addSubview(new WI.CounterView);
        this.addSubview(new WI.CounterView);
        this.addSubview(new WI.CounterView);
    }

    didLayoutSubtree()
    {
        console.log(this.subviews.map((subview) => subview.count).join(" "));
    }
};

let parentView = new WI.ParentView;
WI.View.rootView().addSubview(parentView);
WI.View._visitViewTreeForLayout();

parentView.subviews[1].needsLayout();
parentView.subviews[2].needsLayout();
WI.View._visitViewTreeForLayout();
```

If `didLayoutSubtree` is called after the first child view does its `layout`, then we'd see `1 2 1` get logged.  But the correct output should be `1 2 2` since both the 2nd and the 3rd child views are about to `layout`.

To put it another way and/or (re)summarize, we need a way of knowing "does my parent have any other children that need to `layout`?" in order to correctly call `didLayoutSubtree`.

> Looking at the dirty state of the parent view when doing layout for the subview is moot because it always `false`.

Oh good catch!  I forgot we set it to `false` super early.  We could probably adjust that if needed, but i think your idea of using the `layoutReason` is a bit easier.

> ```
> static _scheduleLayoutForView(view)
> {
> ...
>    view._setDirty(true);
> +  view._setLayoutReason(WI.View.LayoutReason.Dirty);
> ...
> ```

Hmm the only way I can see `layout` getting called without having a `layoutReason` is if the `WI.View` was just added as a subview (which is definitely something we should fix).  All other paths (e.g. `needsLayout`) should already be setting a `layoutReason`.  Is there some other path I'm unaware of?